### DISCUSSION TITLE
Add SSR option to composables

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [nuxt/nuxt#25736] [Ideas]
Hello cool people and nuxters! 🔥💚👋

 Just watched @manniL 's video about using `$fetch` with `useAsyncData` out of setup function in a Repository Pattern, and inspired this feature request.

I propose adding a `ssr: boolean` property to the options of all the Nuxt composables that do SSR logic, so that they can be used in SPA mode and in SSR mode in composables out of Nuxt or Vue context.

(And I have a feeling this would solve the nuxt-auth-utils not working in SPA mode in the process).

### Motivation
I was curiously working with repository pattern these days, because I use NestJS in the backend, and I have controllers, services, modules, database repositories, etc. and want that same design patterns in the frontend.

So I have this structure:

```
~/modules/
    auth/
       auth.controller.ts
       auth.service.ts
       auth.config.ts
       index.ts
     api/
       runtime/
         api.plugin.ts
       index.ts
```

`~/modules/api/runtime/api.plugin.ts`

```ts
import { authController } from '~/modules/auth/auth.config'

export default defineNuxtPlugin(() => {
  return {
    provide: {
       api: {
         auth: authController,
         // Other api here
       }
    }
  }
})
```
And of course, I inject `AuthService` in `AuthController`'s constructor, and `AuthService` is who uses `useAsyncData` with `$fetch` to call the backend API, then I stored the response data in `const user = ref<AuthUser | undefined>()` in `AuthController`, 
and also expose the `error` and `status` refs.

**Note:** 
Tried to type `status` from `useAsyncData` but the `AsyncDataRequestStatus` is not exported... so had to extract it from `AsyncData<AuthUser, NuxtError>['status']['value']`)

**Note:** Had to create `auth.config.ts` to export the `AuthController` instance, because if I export it in index.ts, and then import it in `~/modules/api/runtime/api.plugin.ts` from there, I got an error because it's runtime code that has no access to the `nuxt/kit` package, and index.ts imports stuff from 'nuxt/kit'.

### Final Words
So, it would be nice to tell these composables not to run SSR code, instead of prop drilling the $fetch instance, and with the added benefit of using `useAsyncData` or even `useFetch` inside our repositories/controllers which already contain reactive objects to manage the api response state.

```ts
const {data, error, status} = await $api.auth.login(email.value, password.value)
```

Is better DX than:

```ts
const {data, error, status} = await useAsyncData(() => $api.auth.login(email.value, password.value))
```

**Comment by TheAlexLichter** [comment]
Hey 👋 
Glad that you enjoyed [the video](https://www.youtube.com/watch?v=jXH8Tr-exhI)!

The problem is not SSR vs. SPA. The issue is that **composables** like `useAsyncData`/`useFetch` rely on the Vue and Nuxt context while functions of any kind, like `$fetch` or a `capitalize` utility, do not.


PS: You can tell `useAsyncData`/`useFetch` already to execute on the client-side only via `{ server: false }` as last argument.

**Reply by angelhdzmultimedia** [reply]
Sweet! Thank you so much for the good stuff you're doing! 🔥👏
